src/esockd_acceptor_sup.erl

```diff
@@ -104,10 +107,28 @@ init({AcceptorMod, AcceptorArgs}) ->
 %% Internal functions
 %% -------------------------------------------------------------------
 
-tune_socket_fun(Opts) ->
-    TuneOpts = [ {tune_buffer, proplists:get_bool(tune_buffer, Opts)}
-                 %% optional callback, returns ok | {error, Reason}
-               , {tune_fun, proplists:get_value(tune_fun, Opts, undefined)}],
+tune_socket_fun(Type, Opts) ->
+    Opts1 = case proplists:get_bool(tune_buffer, Opts) of
+                true ->
+                    [{tune_buffer, true}];
+                false ->
+                    []
+            end,
+    Opts2 = case proplists:get_value(tune_fun, Opts) of
+                undefined ->
+                    [];
+                MFA ->
+                    [{tune_fun, MFA}]
+            end,
+    %% if no callbacks provided, run a fast_check to quickly
+    %% fence off invalid sockets such as enotconn and einval
+    %% before spwaning a connection process
+    %% for non-TCP sockets, there are upgrade steps
+    %% to detect errors anyway
+    TuneOpts = case Type =:= tcp of
+                   true -> [fast_check] ++ Opts1 ++ Opts2;
+                   false-> Opts1 ++ Opts2
+               end,
```

Comment:
Nit: the comment suggest that `fast_check` is added only if no callbacks are provided, but only the connection type is checked to decide whether to add `fast_check`.